### PR TITLE
Switch between server and local tracking #12

### PR DESCRIPTION
**Pull request by MichielLangen**
*What changed*
added state pattern for Tracking
added RemoteState for TrackingSession
added unit tests

*What still needs to be done:*
Fix unit tests wich depend on server connentions
add LocalState for TrackingSession

### CODE REVIEW

**Comment by wmeints** [comment]
Also, the `tables` module is missing from the package requirements. You may want to add that one.

File: observatory/tracking.py [hunk 1]
```diff
@@ -1,51 +1,247 @@
...
+
+    @abstractmethod
+    def record_session_end(self, model, version, experiment, run_id, status):
+        """
+        This method will record the end of a run.
+        """
+        pass
+
+class LocalState(ObservatoryState):
+    #Handels metrics for the local filesystem -> Sink.py
```
Comment (wmeints) [inline observatory/tracking.py]: Can you convert this into a proper documentation string?
```suggestion
"""
This state is used to record metadata about experiments in the current working directory using the standard data sink. 
"""
```

File: observatory/tracking.py [hunk 2]
```diff
@@ -1,51 +1,247 @@
...
+    def record_metric(self, name, value):
+        """
+        This method will record a metric.
+        """
+        pass
+
+    @abstractmethod
+    def record_settings(self, name, value):
+        """
+        This method will record a setting.
```
Comment (wmeints) [inline observatory/tracking.py]: I personally feel that comments like this don't help the developer that wants to extend the state mechanism.
Can you change this to include a description like

```suggestion
Override this method in a derived class to record a setting.
The derived class is required to store the settings as a single dictionary per run.
```

Please include a description of the parameters as well.

File: observatory/tracking.py [hunk 3]
```diff
@@ -1,51 +1,247 @@
...
+        """
+        pass
+
+    @abstractmethod
+    def record_output(self, name, value):
+        """
+        This method will record a output.
```
Comment (wmeints) [inline observatory/tracking.py]: I personally feel that comments like this don't help the developer that wants to extend the state mechanism.
Can you change this to include a description like
```suggestion
Override this method in a derived class to record an output for the run.
The derived class is required to handle the value of the output as an opaque binary blob.
It must not read the blob to validate it. 
```

Please include a description of the parameters as well.

File: observatory/tracking.py [hunk 4]
```diff
@@ -1,51 +1,247 @@
...
+        """
+        pass
+
+    @abstractmethod
     def record_session_start(self, model, version, experiment, run_id):
         """
-        Records the start of a session.
+        This method will record the start of a run.
```
Comment (wmeints) [inline observatory/tracking.py]:
```suggestion
Override this method in a derived class to record the start of a session.
The derived class is required to set the status of the run to "In progress." with a unix timestamp.
```

Please include a description of the parameters as well.

File: observatory/tracking.py [hunk 5]
```diff
@@ -1,51 +1,247 @@
...
+        """
+        pass
+
+    @abstractmethod
+    def record_session_end(self, model, version, experiment, run_id, status):
+        """
+        This method will record the end of a run.
```
Comment (wmeints) [inline observatory/tracking.py]:
```suggestion
Override this method in a derived class to record the end of a session.
The derived class must, in addition to the status, also record a unix timestamp.
```

Please include a description of the parameters as well.

File: observatory/tracking.py [hunk 6]
```diff
@@ -1,51 +1,247 @@
...
+
+    def record_session_start(self, model, version, experiment, run_id):
+        print("LocalState : record_session_start")
+
+    def record_session_end(self, model, version, experiment, run_id, status):
+        print("LocalState : record_session_end")
+
+
+class RemoteState(ObservatoryState):
+    #Handels metrics for the remote endpoint
```
Comment (wmeints) [inline observatory/tracking.py]: Can you convert this to a proper documentation string?

```suggestion
"""
Records metric on a remote server that you can run through the command `observatory server`.
"""
```

File: observatory/tracking.py [hunk 7]
```diff
@@ -61,16 +257,21 @@ def record_session_start(self, model, version, experiment, run_id):
...
             The response from the server
         """
-        handler_url = f'{self.url}/api/models/{model}/versions/{version}/experiments/{experiment}/runs'
-        return requests.post(handler_url, json={'run_id': run_id})
+        handler_url = f'{settings.server_url}/api/models/{model}/versions/{version}/experiments/{experiment}/runs/{run_id}/metrics'
+        self._verify_response(requests.post(handler_url, json={'name': name, 'value': value}), 201)
 
-    def record_settings(self, model, version, experiment, run_id, settings):
+    def record_settings(self, model, version, experiment, run_id, name, value):
+        print("RemoteState : record_settings")
```
Comment (wmeints) [inline observatory/tracking.py]: I'm assuming this is going away once we have the remote state fully implemented?

Comment (MichielLangen) [reply observatory/tracking.py]: yes, this one and the print statements in LocalState will be removed during development. They were added to see if the state-pattern worked

**Comment by wmeints** [comment]
👍 Awesome, I'll merge it.

**Pull request merged**